aptos-move/framework/aptos-stdlib/sources/data_structures/ordered_map.move

```diff
@@ -0,0 +1,786 @@
+            e.key
+        })
+    }
+
+    /// Return all values in the map. This requires values to be copyable.
+    public fun values<K, V: copy>(self: &OrderedMap<K, V>): vector<V> {
+        vector::map_ref(&self.entries, |e| {
+            let e: &Entry<K, V> = e;
+            e.value
+        })
+    }
+
+    /// Transform the map into two vectors with the keys and values respectively
+    /// Primarily used to destroy a map
+    public fun to_vec_pair<K, V>(self: OrderedMap<K, V>): (vector<K>, vector<V>) {
+        let keys: vector<K> = vector::empty();
+        let values: vector<V> = vector::empty();
+        let OrderedMap::SortedVectorMap { entries } = self;
+        vector::for_each(entries, |e| {
+            let Entry { key, value } = e;
+            vector::push_back(&mut keys, key);
+            vector::push_back(&mut values, value);
+        });
+        (keys, values)
+    }
+
+    /// For maps that cannot be dropped this is a utility to destroy them
+    /// using lambdas to destroy the individual keys and values.
+    public inline fun destroy<K, V>(
+        self: OrderedMap<K, V>,
+        dk: |K|,
+        dv: |V|
+    ) {
+        let (keys, values) = to_vec_pair(self);
+        vector::destroy(keys, |_k| dk(_k));
+        vector::destroy(values, |_v| dv(_v));
+    }
+
+    /// Apply the function to a reference of each key-value pair in the table.
+    public inline fun for_each_ref<K, V>(self: &OrderedMap<K, V>, f: |&K, &V|) {
+        let iter = self.new_begin_iter();
+        while (!iter.iter_is_end(self)) {
+            f(iter.iter_borrow_key(self), iter.iter_borrow(self));
+            iter = iter.iter_next(self);
+        }
+        // vector::for_each_ref(
+        //     &self.entries,
+        //     |entry| {
+        //         f(&entry.key, &entry.value)
+        //     }
+        // );
+    }
+
+    /// Apply the function to a mutable reference of each key-value pair in the table.
+    public inline fun for_each_mut<K, V>(self: &mut OrderedMap<K, V>, f: |K, &mut V|) {
+        let iter = self.new_begin_iter();
+        while (!iter.iter_is_end(self)) {
+            let key = *iter.iter_borrow_key(self);
+            f(key, iter.iter_borrow_mut(self));
+            iter = iter.iter_next(self);
+        }
+        // vector::for_each_mut(
+        //     &mut self.entries,
+        //     |entry| {
+        //         f(&mut entry.key, &mut entry.value)
+        //     }
+        // );
+    }
+
+    // ========= Section with private methods ===============
+
+    inline fun new_iter(index: u64): Iterator {
+        Iterator::Position {
+            index: index,
+        }
+    }
+
+    // return index containing the key, or insert position.
+    // I.e. index of first element that has key larger or equal to the passed `key` argument.
+    fun binary_search<K, V>(key: &K, entries: &vector<Entry<K, V>>, start: u64, end: u64): u64 {
+        let l = start;
+        let r = end;
```

Comment:
use l and len (r - l) here, and do len/2 every time can probably save a little bit.

Reply:
don't get what you mean, can you clarify? this is also your implementation I copied here, by the way :P

Reply:
Something like this: https://github.com/gcc-mirror/gcc/blob/d9375e490072d1aae73a93949aa158fcd2a27018/libstdc%2B%2B-v3/include/bits/stl_algobase.h#L1023

Reply:
This seems to only make sense where you can have pointers and do pointer arithmetic. Here since you need to access the vector, and always produce index from the start, you need to compute the absolute index anyways - so I don't think that would be faster.